### PR TITLE
feat(codenarc): implement AST-aware diagnostic positioning (Phase 1 POC)

### LINKED ISSUE #674: [diagnostics] Replace heuristic-based positioning with AST-aware positioning
## Problem

Our current `RuleRangeCalculator` uses string-based heuristics to position CodeNarc diagnostics, which is:

1. **Brittle** - Breaks if CodeNarc message formats change
2. **Approximate** - String matching can highlight wrong tokens (e.g., `String field = new String()` highlights the constructor call instead of field name)
3. **Duplicative** - We parse files 2-3 times: once for LSP (GroovyParser), once for CodeNarc, once for string-based positioning
4. **Unmaintainable** - 70+ rules, each with custom regex/string matching logic
5. **Edge case prone** - Fails on nested classes, generics, complex expressions

**Example of current problem:**
```kotlin
// RuleRangeCalculator.kt - Line 528
private fun findLastIdentifier(sourceLine: String): Pair<Int, Int> {
    // Uses regex to find last word - can match wrong token!
    val identifierMatches = Regex("""\b[a-zA-Z_]\w*\b""").findAll(sourceLine).toList()
    // For "String field = new String()" this returns "String" at end (wrong!)
}
```

## Context

Discovered during PR #666 review when fixing diagnostic positioning bugs. We realized we're using heuristics when we already have a fully parsed AST with exact position information available from `GroovyParser`.

**We already have the infrastructure:**
- ✅ `ClassDeclaration`, `MethodDeclaration`, `FieldDeclaration` AST nodes
- ✅ Each node has `range: Range?` with exact `begin/end` positions  
- ✅ `CoordinateSystem` to convert Groovy (1-based) ↔ LSP (0-based)
- ✅ Parser already used by LSP for highlighting, completion, navigation

**Current architecture duplicates work:**
```
File saved
  ↓
  ├→ [LSP] Parse with GroovyParser → AST₁ → syntax highlighting, completion
  ├→ [CodeNarc] Parse with Groovy compiler → AST₂ → violations (line number only)
  └→ [RuleRangeCalculator] String regex on sourceLine → column position (heuristic)
```

### **Phase 1: Proof of Concept** (POC)

Implement AST-based positioning for 5 high-value rules to validate the approach:
- `ClassName`, `MethodName`, `FieldName`, `UnusedVariable`, `UnusedImport`

**New classes:**
```kotlin
// 1. Find AST nodes by line number and rule type
class AstNodeFinder(private val ast: CompilationUnit) {
    fun findNodeAtLine(line: Int, ruleType: String): Node?
}

// 2. Extract precise ranges from AST nodes
class AstAwareRangeCalculator(private val coordinateSystem: CoordinateSystem) {
    fun calculateRangeFromAst(node: Node, violation: Violation): Pair<Int, Int>?
}

// 3. Hybrid system: AST first, fallback to heuristics
class HybridRangeCalculator(
    private val astCalculator: AstAwareRangeCalculator,
    private val heuristicCalculator: RuleRangeCalculator
) {
    fun calculateRange(
        violation: Violation, 
        ast: CompilationUnit?, 
        sourceLine: String
    ): Pair<Int, Int>
}
```

**Measure:**
- Accuracy improvement (% of correct highlights vs heuristic)
- Performance impact (with/without AST caching)
- Code complexity reduction

### **Phase 2: Hybrid System**

Categorize rules and use AST for structural violations, keep heuristics for formatting:

**AST-based (deterministic) - ~60% of rules:**
- Naming: `ClassName`, `MethodName`, `FieldName`, `VariableName`, `ParameterName`
- Unused: `UnusedVariable`, `UnusedPrivateField`, `UnusedPrivateMethod`
- Imports: `UnusedImport`, `DuplicateImport`
- Empty: `EmptyMethod`, `EmptyClass`, `EmptyIfStatement`
- Exceptions: `CatchException`, `ThrowException`

**Heuristic-based (no AST node) - ~40% of rules:**
- Formatting: `TrailingWhitespace`, `Indentation`, `LineLength`
- Spacing: `SpaceAfterComma`, `SpaceBeforeOpeningBrace`
- Syntax: `UnnecessarySemicolon`

### **Phase 3: Integration & Optimization**

Integrate with `CodeNarcDiagnosticProvider` and optimize performance:

```kotlin
class AstAwareDiagnosticProvider(
    private val parser: GroovyParser,
    private val codeAnalyzer: CodeAnalyzer
) {
    fun analyzeAndGetDiagnostics(sourceCode: String, fileName: String): List<Diagnostic> {
        // 1. Parse once with our parser (or reuse LSP cache)
        val ast = parser.parse(sourceCode).result.orElse(null)
        
        // 2. Run CodeNarc analysis
        val results = codeAnalyzer.analyze(...)
        
        // 3. Convert violations using AST for positioning
        return results.violations.map { violation ->
            val range = hybridCalculator.calculateRange(violation, ast, sourceLine)
            createDiagnostic(violation, range)
        }
    }
}
```

**Optimizations:**
- Cache parsed ASTs (reuse LSP parse)
- Lazy AST lookup (only for rules that need it)
- Share AST between LSP and diagnostics

### **Phase 4: Migration & Cleanup**

- Migrate all 70+ structural rules to AST-based
- Remove redundant string matching code from `RuleRangeCalculator`
- Keep only formatting rules as heuristics
- Performance benchmarks and regression tests

## Benefits

1. **Deterministic** - Direct from AST, no guessing
2. **Accurate** - Zero false positives from string matching
3. **Maintainable** - Add new rules by mapping to node types, not writing regex
4. **Testable** - Test with AST fixtures instead of string patterns
5. **Robust** - Handles edge cases automatically (nested classes, generics, complex expressions)
6. **Performance** - Potentially faster with caching (single parse instead of multiple)
7. **Future-proof** - Easy to extend for new CodeNarc rules

## Implementation Checklist

- [ ] Phase 1: POC with 5 rules
  - [ ] Implement `AstNodeFinder`
  - [ ] Implement `AstAwareRangeCalculator`
  - [ ] Implement `HybridRangeCalculator`
  - [ ] Battle tests comparing accuracy
  - [ ] Performance benchmarks
- [ ] Phase 2: Hybrid system
  - [ ] Categorize all CodeNarc rules (AST vs heuristic)
  - [ ] Migrate structural rules
  - [ ] Keep formatting rules as heuristics
- [ ] Phase 3: Integration
  - [ ] Integrate with `CodeNarcDiagnosticProvider`
  - [ ] Add AST caching
  - [ ] Share parsed ASTs with LSP
- [ ] Phase 4: Cleanup
  - [ ] Remove unused heuristic code
  - [ ] Update all tests
  - [ ] Documentation

## References

- Current implementation: `groovy-diagnostics/codenarc/src/main/kotlin/.../RuleRangeCalculator.kt`
- AST infrastructure: `parser/core/src/main/kotlin/.../ast/`
- Coordinate system: `parser/native/src/main/kotlin/.../CoordinateSystem.kt`
- PR #666 where this was discovered
- Related bug fixes in commits 3ab130b6, ab505364

### PR DESCRIPTION
**Pull request by albertocavalcante**
## Summary

- Implement AST-aware diagnostic positioning for 5 CodeNarc rules as a proof-of-concept
- Replace heuristic-based string matching with precise AST-based identifier location
- Introduce hybrid approach that falls back to heuristics when AST is unavailable

## Phase 1 POC Rules

1. **ClassName** - highlight class name identifier
2. **MethodName** - highlight method name identifier  
3. **FieldName** - highlight field name identifier
4. **UnusedVariable** - highlight variable name
5. **UnusedImport** - highlight import class name

## Architecture

```
CompilationAccessor (interface)
    └── Provides cached ModuleNode from LSP compilation service

AstNodeFinder
    └── Locates AST nodes by line number from native Groovy AST
    └── findClassAtLine(), findMethodAtLine(), findFieldAtLine()
    └── findVariableAtLine(), findImportAtLine()

AstAwareRangeCalculator  
    └── Extracts precise ranges from AST + source line
    └── Uses word boundary regex to find exact identifier position

HybridRangeCalculator
    └── Delegates to AST when available and supported
    └── Falls back to proven heuristic approach otherwise
```

## Benefits

- **More accurate positioning**: AST-based lookup finds exact identifier location regardless of modifiers, annotations, or keywords on the line
- **No regression risk**: Hybrid approach ensures fallback to existing heuristic behavior
- **Minimal integration impact**: Uses existing cached AST from `GroovyCompilationService`
- **Extensible**: Foundation for Phase 2+ to support more rules

## Test plan

- [ ] All 47 new unit tests pass (`AstNodeFinderTest`, `AstAwareRangeCalculatorTest`, `HybridRangeCalculatorTest`, `AstVsHeuristicAccuracyTest`)
- [ ] Existing `DiagnosticPositioningBattleTest` passes (no regression)
- [ ] Full `groovy-diagnostics:codenarc:test` suite passes
- [ ] Manual testing: Open VS Code with Groovy file, trigger violations for the 5 POC rules, verify correct highlighting

Closes #674

---

> [!NOTE]
> Introduces precise, AST-backed ranges for CodeNarc diagnostics while preserving existing behavior via a hybrid fallback.
> 
> - New `CompilationAccessor` API to retrieve cached `ModuleNode`; adds `groovy.core` dependency
> - Implements `AstNodeFinder` and `AstAwareRangeCalculator` for AST-based positioning; supports `ClassName`, `MethodName`, `FieldName`, `UnusedVariable`, `UnusedImport`
> - Adds `HybridRangeCalculator` to prefer AST when available, otherwise use `RuleRangeCalculator`
> - Updates `CodeNarcDiagnosticProvider` to wire in `CompilationAccessor`, pass `uri`, and use `HybridRangeCalculator` during violation-to-diagnostic conversion
> - Extensive unit tests covering AST lookup, hybrid behavior, and AST vs heuristic accuracy
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 1f2576cc74b3fe8350ce86057c87ba6520be717c. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

## Summary by CodeRabbit

* **New Features**
  * AST-aware diagnostic positioning for more precise violation highlighting (class, method, field, unused variable, unused import).
  * Hybrid positioning that falls back to heuristics when AST data is unavailable.
  * Support for a shared, cached AST accessible across features.

* **Tests**
  * Added comprehensive tests covering AST lookups, AST vs heuristic accuracy, and hybrid behavior.

* **Chores**
  * Added Groovy core dependency to the build.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: groovy-diagnostics/codenarc/src/main/kotlin/com/github/albertocavalcante/diagnostics/codenarc/AstNodeFinder.kt [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+        if (expression.lineNumber == targetLine) {
+            // variableExpression can be null for tuple declarations like "def (x, y) = [1, 2]"
+            val variable = expression.variableExpression
+            if (variable != null && variable.name == targetName) {
+                result = variable
+                return
+            }
+        }
+        super.visitDeclarationExpression(expression)
+    }
```
[the remark this answers is not in the document]
Comment (albertocavalcante) [reply groovy-diagnostics/codenarc/src/main/kotlin/com/github/albertocavalcante/diagnostics/codenarc/AstNodeFinder.kt]: Fixed in 741199af - added TupleExpression support for tuple declarations like `def (x, y) = [1, 2]`.

File: groovy-diagnostics/codenarc/src/main/kotlin/com/github/albertocavalcante/diagnostics/codenarc/AstNodeFinder.kt [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+        for (classNode in moduleNode.classes) {
+            for (method in classNode.methods) {
+                method.code?.visit(visitor)
+                if (visitor.result != null) {
+                    return visitor.result
+                }
+            }
+        }
+        return null
+    }
```
[the remark this answers is not in the document]
Comment (albertocavalcante) [reply groovy-diagnostics/codenarc/src/main/kotlin/com/github/albertocavalcante/diagnostics/codenarc/AstNodeFinder.kt]: Fixed in 741199af - added recursive inner class search to findVariableAtLine for consistency with other finder methods.

File: groovy-diagnostics/codenarc/src/main/kotlin/com/github/albertocavalcante/diagnostics/codenarc/AstNodeFinder.kt [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+            }
+            // Search in inner classes
+            val innerClasses = classNode.innerClasses?.asSequence()?.toList() ?: emptyList()
+            val found = findFieldAtLineRecursive(innerClasses, groovyLine)
+            if (found != null) {
+                return found
+            }
+        }
+        return null
+    }
```
[the remark this answers is not in the document]
Comment (albertocavalcante) [reply groovy-diagnostics/codenarc/src/main/kotlin/com/github/albertocavalcante/diagnostics/codenarc/AstNodeFinder.kt]: Created #723 to track DRY refactoring. Out of scope for this POC PR - stylistic improvement that can be addressed separately.

File: groovy-diagnostics/codenarc/src/main/kotlin/com/github/albertocavalcante/diagnostics/codenarc/AstNodeFinder.kt [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    fun findClassAtLine(groovyLine: Int): ClassNode? = findClassAtLineRecursive(moduleNode.classes, groovyLine)
+
+    private fun findClassAtLineRecursive(classes: List<ClassNode>, groovyLine: Int): ClassNode? {
+        for (classNode in classes) {
+            // Check if this class is declared at the target line
+            if (classNode.lineNumber == groovyLine) {
+                return classNode
+            }
+            // Check inner classes recursively
+            val innerClasses = classNode.innerClasses?.asSequence()?.toList() ?: emptyList()
```
[the remark this answers is not in the document]
Comment (albertocavalcante) [reply groovy-diagnostics/codenarc/src/main/kotlin/com/github/albertocavalcante/diagnostics/codenarc/AstNodeFinder.kt]: Investigated in 741199af - `asSequence()` is actually required here. `classNode.innerClasses` returns an `Iterator<InnerClassNode>`, not a `List`. In Kotlin, `Iterator` doesn't have a `toList()` extension function directly - you need `asSequence().toList()` to convert. Keeping as-is.
